### ISSUE TITLE
Firewalld rich rules and/or service support for ipsec match policy

### ISSUE THREAD

**Issue by fpqc** [t-woerner/firewalld#163]
So with iptables, you might want to add this rule to allow an ipsec daemon to forward traffic onwards to the LAN:

`iptables -t filter -A FORWARD -o eth1 -m policy --dir in --pol ipsec -j ACCEPT`

This is equivalent to the --direct rule

`firewall-cmd --direct --add-rule ipv4 filter FORWARD 0 -o eth1 -m policy --dir in --pol ipsec -j ACCEPT`

But if you look at that rule, there's no real reason why you shouldn't be able to add this as a service.  It applies a policy to a particular interface, which belongs to a zone (in my case, the 'work' zone).  

The trouble is that there's no way (as far as I know) to add a 'match' (`-m`) rule to a firewall-cmd service.  It also happens to be the case that there is no reason why one would need to write this rule, which doesn't directly refer to IP addresses, separately for ipv4 and ipv6. 

Will we maybe see support for this kind of rule directly in firewalld in the future?

**Comment by rocketraman** [comment]
I migrated from shorewall to firewalld and it wasn't obvious how to set up my firewall to support ipsec. With shorewall, I just configured an entry in `tunnels`, an `ipsec` zone in `zones`, a routing entry in `hosts`, and the appropriate policies.

After migrating to firewalld, it appears as if the setup is completely manual. For posterity, here is what I did (my ipsec tunnel is provided by strongswan, and `10.0.0.0/8` is my remote network):

```
firewall-cmd --direct --add-rule ipv4 filter FORWARD 0 -d 10.0.0.0/8 -m policy --dir out --pol ipsec --mode tunnel -j ACCEPT
firewall-cmd --direct --add-rule ipv4 filter FORWARD 1 -s 10.0.0.0/8 -m policy --dir in --pol ipsec --mode tunnel -j ACCEPT
firewall-cmd --direct --add-rule ipv4 nat POSTROUTING 0 -m policy --dir out --pol ipsec -j ACCEPT
```

**Comment by fpqc** [comment]
@rocketraman why are you forwarding traffic for strongswan like this? My policy was for the gateway side, not client-side.  You shouldn't need to set firewall rules for strongswan as a client iirc.

**Comment by rocketraman** [comment]
> @rocketraman why are you forwarding traffic for strongswan like this

@fpqc  You're right, the forwarding rules were unnecessary -- I just removed them and everything still works. However the `POSTROUTING` rule still seems necessary. Related docs here:

https://wiki.strongswan.org/projects/strongswan/wiki/ForwardingAndSplitTunneling#General-NAT-problems

**Comment by rocketraman** [comment]
Also: I'd still like to be able to specify firewall rules for ipsec traffic inbound by setting a different zone for that traffic, just like I was able to do with shorewall. Still not sure how to do that...

**Comment by tahiro86j** [comment]
Previously I allowed access to any services on the debian machine (VPS, with a global IPv4) given that the intended source address is static, hence the address of Cisco ISR. I added those addresses to the "trusted" zone. I've recently established transport sessions between a Cisco ISR and strongswan on debian. Reason: I now need to run some DNS updates over UDP and also do AXFR as needed.
What seems smart to me is if I could regard any traffic via ipsec as that of the trusted zone. How can I do this?
